File: in_ruby/day_01_part_2.rb
input = File.read("../inputs/day01.txt").lines.map(&:to_i)

mass_to_fuel = -> (mass) {
  fuel_mass = mass.div(3) - 2
  if fuel_mass > 0
    fuel_mass + mass_to_fuel[fuel_mass]
  else
    0
  end
}

fuel = input.map(&mass_to_fuel).reduce(&:+)

#--
puts "Day 01 - Part 2: #{fuel}"

File: in_haskell/day_01_part_2.hs
#!/usr/bin/env stack --resolver lts-14.14 runghc

parseInt :: String -> Int
parseInt x = read $ filter (`elem` "-0123456789") x

loadInput :: IO ([Int])
loadInput = do
  raw <- readFile "../inputs/day01.txt"
  return $ map parseInt $ lines raw

massToFuel :: Int -> Int
massToFuel m = if fuelMass > 0
               then fuelMass + (massToFuel fuelMass)
               else 0
               where fuelMass = (m `div` 3) - 2

main :: IO ()
main = do
  input <- loadInput
  putStr "Day 01 - Part 2: "
  putStrLn $ show $ foldl (+) 0 $ map massToFuel input
